### DISCUSSION TITLE
How to get the points a, b of the segment shape?

### DISCUSSION THREAD

**Discussion by setanarut** [jakecoffman/cp#29] [Q&A] [answered]
## How to get the points a, b of the segment shape?

`cp.NewSegment()` returns a `*cp.Shape` struct. When I create `&cp.Segment{}` separately, `Segment.A()` returns zero values. 

`Shape.Point(0)` returns `cp.SupportPoint`. A completely useless structure (no exported fields)

example code:

```Go
package main

import (
	"fmt"

	"github.com/jakecoffman/cp/v2"
)

func main() {
	space := cp.NewSpace()
	space.SetGravity(cp.Vector{X: 0, Y: -20})
	space.Iterations = 10

	seg := &cp.Segment{}
	body := cp.NewStaticBody()
	seg.Shape = cp.NewSegment(
		body,
		cp.Vector{X: 50, Y: 10},
		cp.Vector{X: 350, Y: 13},
		0)
	body.SetPosition(cp.Vector{10, 10})

	space.AddBody(body)
	space.AddShape(seg.Shape)

	space.Step(1. / 60.)

	fmt.Printf("%T %T \n", seg.Shape.Point(0), seg.Shape.Point(1))
	fmt.Printf("%v %v \n", seg.Shape.Point(0), seg.Shape.Point(1))
	fmt.Printf("%v %v \n", seg.A(), seg.B())
	
	// Output:

	// cp.SupportPoint cp.SupportPoint 
	// {{60 20} 0} {{360 23} 1} 
	// 0.000000,0.000000 0.000000,0.000000 
```

---

I think I need this function. `cpSegmentShapeGetA(shape)` :
https://github.com/slembcke/Chipmunk2D/blob/d0239ef4599b3688a5a336373f7d0a68426414ba/src/cpShape.c#L511

**Comment by jakecoffman** [comment, marked as answer]
Right, it gets weird if you initialize structs directly. I would do it like this:

```go

func main() {
	space := cp.NewSpace()
	space.SetGravity(cp.Vector{X: 0, Y: -20})
	space.Iterations = 10

	body := cp.NewStaticBody()
	segShape := cp.NewSegment(
		body,
		cp.Vector{X: 50, Y: 10},
		cp.Vector{X: 350, Y: 13},
		0)
	body.SetPosition(cp.Vector{10, 10})

	space.AddBody(body)
	space.AddShape(segShape)

	space.Step(1. / 60.)

	fmt.Printf("%T %T \n", segShape.Point(0), segShape.Point(1))
	fmt.Printf("%v %v \n", segShape.Point(0), segShape.Point(1))
	seg := segShape.Class.(*cp.Segment)
	fmt.Printf("%v %v \n", seg.A(), seg.B())
}
```

**Reply by setanarut** [reply]
Thanks. I tried I think `segShape.Class.(*cp.Segment).TransformA()` gives the modified position. this is what I need. 

Also I think it should be like this to make more sense:

Rename `NewSegment()` to `NewSegmentShape()` (misleading name)

then:

`SegmenShapeGetA(segShape)`

or 

`segShape.A()`

**Comment by setanarut** [comment]
Here are the missing functions:

https://chipmunk-physics.net/release/ChipmunkLatest-Docs/#cpShape-Segments
![Ekran Resmi 2024-01-06 01 34 02](https://github.com/jakecoffman/cp/assets/36481442/c453cc01-d447-47ff-893a-8ea4e70e3d69)
